src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocumentState.cs

```diff
@@ -191,12 +191,27 @@ public SourceGeneratedDocumentState WithGenerationDateTime(DateTime generationDa
 
     public SourceGeneratedDocumentState WithSyntaxRoot(SyntaxNode newRoot)
     {
-        // See if we can reuse this instance directly
-        if (this.TryGetSyntaxTree(out var tree) && tree == newRoot.SyntaxTree)
-            return this;
+        SyntaxTree newTree;
+        if (this.TryGetSyntaxTree(out var tree))
+        {
+            // See if we can reuse this instance directly
+            if (tree.TryGetRoot(out var root) && root == newRoot)
+                return this;
+
+            newTree = tree.WithRootAndOptions(newRoot, ParseOptions);
+        }
+        else
+        {
+            // We don't have a tree yet, so we need to create one. Using "newRoot.SyntaxTree" is the last resort though
+            // as it could lazily create one, which doesn't know about our parse options.
+            var factory = LanguageServices.GetService<ISyntaxTreeFactoryService>();
+            newTree = factory is not null
+                ? factory.CreateSyntaxTree(Attributes.SyntaxTreeFilePath, ParseOptions, SourceText, SourceText.Encoding, SourceText.ChecksumAlgorithm, newRoot)
+                : newRoot.SyntaxTree;
```

Comment:
It's not clear to me why we need this case -- can't we assume we have a ISyntaxTreeFactoryService, because without one we never could have parsed a source generated document in the first place?

Reply:
If you say we can, I'm happy to enforce that in the code, I'm just being extra cautious with services I don't understand :)